src/rntuple/RAMNTupleRecord.cxx

```diff
@@ -0,0 +1,749 @@
+   if (!file.IsOpen())
+      return;
+   file.cd();
+
+   // Create a simple model for metadata
+   auto metaModel = RNTupleModel::Create();
+   auto rnameField = metaModel->MakeField<std::vector<std::string>>("rname_refs");
+   auto rnextField = metaModel->MakeField<std::vector<std::string>>("rnext_refs");
+
+   ROOT::RNTupleWriteOptions writeOptions;
+   writeOptions.SetCompression(505);
+
+   auto metaWriter = RNTupleWriter::Append(std::move(metaModel), "METADATA", file, writeOptions);
+   auto metaEntry = metaWriter->CreateEntry();
+   auto rnamePtr = metaEntry->GetPtr<std::vector<std::string>>("rname_refs");
+   auto rnextPtr = metaEntry->GetPtr<std::vector<std::string>>("rnext_refs");
+
+   *rnamePtr = fgRnameRefs->GetRefs();
+   *rnextPtr = fgRnextRefs->GetRefs();
+   metaWriter->Fill(*metaEntry);
+}
+
+void RAMNTupleRecord::ReadAllRefs(const std::string &filename)
+{
+   try {
+      auto reader = RNTupleReader::Open("METADATA", filename);
+      if (!reader || reader->GetNEntries() == 0)
+         return;
+
+      try {
+         auto refs_view = reader->GetView<std::vector<std::string>>("rname_refs");
+         const auto &refs = refs_view(0);
+         fgRnameRefs->Clear();
+         for (const auto &ref : refs) {
+            fgRnameRefs->AddRef(ref);
+         }
+      } catch (...) {
+         // Field doesn't exist
+      }
+
+      // Read next reference names
+      try {
+         auto rnext_view = reader->GetView<std::vector<std::string>>("rnext_refs");
+         const auto &refs = rnext_view(0);
+         fgRnextRefs->Clear();
+         for (const auto &ref : refs) {
+            fgRnextRefs->AddRef(ref);
+         }
+      } catch (...) {
+         // Field doesn't exist
+      }
+   } catch (...) {
+      // Metadata might not exist
+   }
+}
+
+void RAMNTupleRecord::WriteIndex(TFile &file)
+{
+   if (!fgIndex || fgIndex->Size() == 0 || !file.IsOpen())
+      return;
+   file.cd();
+
+   // Create index model
+   auto indexModel = RNTupleModel::Create();
+   auto indexField = indexModel->MakeField<std::vector<RAMNTupleIndex::IndexEntry>>("index_entries");
+
+   ROOT::RNTupleWriteOptions writeOptions;
+   writeOptions.SetCompression(505);
+
+   auto indexWriter = RNTupleWriter::Append(std::move(indexModel), "INDEX", file, writeOptions);
+   auto indexEntry = indexWriter->CreateEntry();
+   auto indexPtr = indexEntry->GetPtr<std::vector<RAMNTupleIndex::IndexEntry>>("index_entries");
+
+   *indexPtr = fgIndex->GetEntries();
+   indexWriter->Fill(*indexEntry);
+}
+
+void RAMNTupleRecord::ReadIndex(const std::string &filename)
+{
+   try {
+      auto reader = RNTupleReader::Open("INDEX", filename);
+      if (!reader || reader->GetNEntries() == 0)
+         return;
+
+      try {
+         auto index_view = reader->GetView<std::vector<RAMNTupleIndex::IndexEntry>>("index_entries");
+         const auto &entries = index_view(0);
+         fgIndex->SetEntries(entries);
+      } catch (...) {
+         // Field doesn't exist
```

Comment:
That's something we should rework maybe not part of this PR. We should not rely on exceptions to check if a field exists or not...